src/soap.js

```diff
@@ -543,6 +545,29 @@ class SoapMethodCall {
         return options;
     }
 
+    /**
+     * Re-Finalize a SOAP call just before sending
+     * @param {object} client the ACC js sdk client
+     */
+    reFinalize(client) {
+        const sessionToken  = client._sessionToken;
+        const securityToken = client._securityToken;
+        this._sessionToken = sessionToken;
+        var cookie = DomUtil.findElement(this._header, "Cookie");
+        if (!cookie) {
+            cookie = this._doc.createElement("Cookie");
+            this._header.appendChild(cookie);
+        }
+        cookie.textContent = `__sessiontoken=${this._sessionToken}`;
+        // MethodSession element exit always
+        var methodSession = DomUtil.findElement(this._method, "sessiontoken");
+        methodSession.textContent = this._sessionToken;
+        this._securityToken = securityToken
+        var security = DomUtil.findElement(this._header, "X-Security-Token");
+        security.textContent = securityToken;
+        this.finalize(this.request.url);
+    }
+    
```

Comment:
I'm wondering about this reFinalize method, I'm afraid it's going to be brittle because there's a strong correlation with init method. For instance, in init, we use client._sessionToken || "", but we do not add the || "" in reFinalize. Why? Is it important?

Also, the Cookie element may be added at the wrong place: in refinalize, it's added at the end of the document, whereas it's normally added at the beginning of the document. Again, is it important?

Instead, I'm wondering if we could modify as follow:
* Modify the "_initMessage" function to remove all the code related to setting the cookie, session and security token
* Instead put this code in the "finalize" function, i.e. finalize will take the "client" as first argument
* replace reFinalize by a call to finalize. In other words finalize can be called multiple times and does 2 things now : set the auth tokens, and build the http request.

Reply:
code refacto done.